tests/unit/staff/views/test_projects.py

```diff
@@ -390,6 +390,7 @@ def test_projectedit_post_success(rf, core_developer):
     assert set_from_qs(project.orgs.all()) == {new_org.pk}
     assert project.updated_by == core_developer
     assert project.redirects.count() == 1
+    assert data["slug"] not in project.redirects.first().old_url
```

Comment:
I found it really hard to understand the intention of this assertion. For sure, this is mostly because of the pre-existing code: `data` is an unhelpful name; something like `updated_project_data` would be better. And `old_url` is similarly unhelpful. I think of a redirect as being _from_ something _to_ something. Mentally replacing _old_ with _from_ is cognitively demanding. However, it's in our power to refactor the pre-existing code. Renaming `data` to `updated_project_data` is scoped to the test function. We should consider it.

That said, I still think it's really hard to understand the intention of this assertion. The issue (#4158) concerns a circular redirect; from a slug to the same slug. If we saved the slug, we could use it in the assertion. Something like the following in the _arrange_ step of the test:

```py
old_project_slug = project.slug
```

And something like the following in the _assert_ step of the test:

```py
assert old_project_slug in project.redirects.first().old_url, "Circular redirect"
```

Notice that we're now testing that an expected outcome is the case -- "The redirect should contain the old project slug"[^1] -- and if it isn't the case, then we're supplying a human-readable message.

[^1]: Actually, "The old project slug is contained by the redirect". Python could be more helpful, here.

Reply:
> `assert old_project_slug in project.redirects.first().old_url, "Circular redirect"`

I think this is indequate in the case where a slug changes from "something" to "something-123". In the circular redirect case the `old_url` would be "something-123" which also contains "something".

Similarly my initial code both is also inadequate for "something-123" becoming "something".

I think a truly robust solution needs to extract the slug from the Redirect `old_url` (as the Redirect doesn't contain the raw slug) and the compare the slugs exactly.